substratevm/src/com.oracle.svm.reflect/src/com/oracle/svm/reflect/serialize/hosted/SerializationFeature.java

```diff
@@ -40,82 +40,170 @@
 import java.net.URL;
 import java.net.URLClassLoader;
 import java.security.NoSuchAlgorithmException;
+import java.util.Collection;
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.HashSet;
-import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
 
 import org.graalvm.nativeimage.ImageSingletons;
 import org.graalvm.nativeimage.hosted.Feature;
 import org.graalvm.nativeimage.hosted.RuntimeReflection;
+import org.graalvm.nativeimage.impl.RuntimeSerializationSupport;
 
 import com.oracle.svm.core.annotate.AutomaticFeature;
 import com.oracle.svm.core.configure.ConfigurationFiles;
 import com.oracle.svm.core.configure.SerializationConfigurationParser;
-import com.oracle.svm.core.configure.SerializationConfigurationParser.SerializationParserFunction;
+import com.oracle.svm.core.configure.SerializationDenyConfigurationParser;
 import com.oracle.svm.core.jdk.Package_jdk_internal_reflect;
-import com.oracle.svm.core.jdk.serialize.SerializationRegistry;
 import com.oracle.svm.core.option.SubstrateOptionsParser;
 import com.oracle.svm.core.util.UserError;
 import com.oracle.svm.core.util.VMError;
-import com.oracle.svm.core.util.json.JSONParserException;
 import com.oracle.svm.hosted.FallbackFeature;
 import com.oracle.svm.hosted.FeatureImpl;
 import com.oracle.svm.hosted.ImageClassLoader;
 import com.oracle.svm.hosted.NativeImageOptions;
 import com.oracle.svm.hosted.config.ConfigurationParserUtils;
+import com.oracle.svm.hosted.config.SerializationRegistryAdapter;
 import com.oracle.svm.reflect.serialize.SerializationSupport;
 import com.oracle.svm.util.ReflectionUtil;
 import com.oracle.svm.util.SerializationChecksumCalculator;
 
-import jdk.vm.ci.meta.MetaUtil;
-
 @AutomaticFeature
 public class SerializationFeature implements Feature {
+    private RuntimeSerializationSupportImpl runtimeSerializationSupport;
     private int loadedConfigurations;
 
     @Override
-    public void beforeAnalysis(BeforeAnalysisAccess a) {
-        FeatureImpl.BeforeAnalysisAccessImpl access = (FeatureImpl.BeforeAnalysisAccessImpl) a;
+    public void duringSetup(DuringSetupAccess a) {
+        FeatureImpl.DuringSetupAccessImpl access = (FeatureImpl.DuringSetupAccessImpl) a;
+
+        SerializationDenyRegistry serializationDenyRegistry = new SerializationDenyRegistry();
         SerializationBuilder serializationBuilder = new SerializationBuilder(access);
+        runtimeSerializationSupport = new RuntimeSerializationSupportImpl(serializationDenyRegistry, serializationBuilder);
```

Comment:
`RuntimeSerializationSupportImpl` is in package `com.oracle.svm.reflect.serialize.hosted` yet it extends `SerializationSupport`. We do not allow classes with package names `*.hosted` to be part of the image heap. But this is exactly what happens when you make the `RuntimeSerializationSupport` ImageSingleton a `*.hosted`-class. We depend on the `RuntimeSerializationSupport` ImageSingleton at image runtime (via `getSerializationConstructorAccessor`). Please keep classes that end up in the image heap strictly separate from classes that are only needed at image build time (for configuration).

Reply:
I suggest you instead make SerializationBuilder extend `RuntimeSerializationSupport`. Then `RuntimeSerializationSupport` could be an additional ImageSingleton that only exists at image built time and does not have anything to do with the `SerializationRegistry` ImageSingleton that is needed in the image heap.

Reply:
Does it make sense to keep the content of `RuntimeSerializationSupportImpl` separate from its parent `SerializationSupport`? Looking at the way `ReflectionDataBuilder` is implemented for reflection support, it would probably make sense to simply merge these 2 in only the `SerializationSupport` class.

Reply:
I didn't see the second comment before commenting my self on this, so appologies for the notification spam probably.

There is no logic provided for ImageSingleton to remove a singleton. Am I right in assuming that, since ImageSingleton lookups are done at build-time, the additional `SerializationBuilder` stored in the ImageSingletons will be automatically removed since it's no longer reachable?

Reply:
> the additional SerializationBuilder stored in the ImageSingletons will be automatically removed since it's no longer reachable?

Correct!

Reply:
fixed